File: templates/templates/typescriptv2/tests/security-options/auth_additional.test.ts
import { expect, test } from "vitest";

import { SDK } from "../index.js";
import { SDKError } from "../sdk/models/errors/sdkerror.js";
import { recordTest } from "./common_helpers.js";

test("Global Security Basic Http Success", async () => {
  recordTest("auth-basic-http-global-option");

  // Expected to succeed since BasicHTTP takes priority over AccessToken in global security definition
  const sdk = new SDK({
    security: {
      basicHttp: {
        username: "testUser",
        password: "testPass",
      },
      accessToken: {
        accessToken: "Bearer ignored",
      },
    },
  });

  const res = await sdk.auth.globalSecurityOptionBasicHttp();

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.basicAuth).toBeDefined();
  expect(res.basicAuth!.authenticated).toBe(true);
  expect(res.basicAuth!.user).toBe("testUser");
});

test("Global Security Fields Ordering", async () => {
  recordTest("auth-global-security-option-fields-ordering");

  // Expected to fail since APIKeyAuth takes priority over BasicHTTP in global security definition
  const sdk = new SDK({
    security: {
      apiKeyAuth: {
        apiKeyAuth: "Bearer test_api_key",
      },
      basicHttp: {
        username: "testUser",
        password: "testPass",
      },
    },
  });

  try {
    await sdk.auth.globalSecurityOptionBasicHttp();
    expect.unreachable("Expected request to fail with 401");
  } catch (e) {
    expect(e).toBeInstanceOf(SDKError);
    expect((e as SDKError).httpMeta.response.status).toBe(401);
  }
});

test("Hoisted Security Access Token First", async () => {
  recordTest("auth-hoisted-security-option-access-token-first");

  const sdk = new SDK({
    security: {
      accessToken: {
        accessToken: "Bearer ghp_xxxx",
      },
    },
  });

  const res = await sdk.auth.hoistedSecurityOptionAccessTokenFirst();

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.tokenAuthResponse).toBeDefined();
  expect(res.tokenAuthResponse!.token).toBe("Bearer ghp_xxxx");
});

test("Hoisted Security Api Key First", async () => {
  recordTest("auth-hoisted-security-option-api-key-first");

  const sdk = new SDK({
    security: {
      apiKeyAuth: {
        apiKeyAuth: "testApiKey",
      },
    },
  });

  const res = await sdk.auth.hoistedSecurityOptionApiKeyFirst();

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.tokenAuthResponse).toBeDefined();
  expect(res.tokenAuthResponse!.token).toBe("testApiKey");
});

test("Hoisted Security Basic Http Only", async () => {
  recordTest("auth-hoisted-security-option-basic-http-only");

  const sdk = new SDK({
    security: {
      basicHttp: {
        username: "testUser",
        password: "testPass",
      },
    },
  });

  const res = await sdk.auth.hoistedSecurityOptionBasicHttpOnly();

  expect(res.httpMeta.response.status).toBe(200);
  expect(res.basicAuth).toBeDefined();
  expect(res.basicAuth!.authenticated).toBe(true);
  expect(res.basicAuth!.user).toBe("testUser");
});

test("Hoisted Security Invalid Option", async () => {
  recordTest("auth-hoisted-security-invalid-option");

  // Provide only basicHttp — not valid for accessTokenFirst which expects bearer/apiKey
  const sdk = new SDK({
    security: {
      basicHttp: {
        username: "user",
        password: "pass",
      },
    },
  });

  try {
    await sdk.auth.hoistedSecurityOptionAccessTokenFirst();
    expect.unreachable("Expected request to fail with 401");
  } catch (e) {
    expect(e).toBeInstanceOf(SDKError);
    expect((e as SDKError).httpMeta.response.status).toBe(401);
  }
});

File: templates/templates/javav2/tests/security-options/AuthAdditionalTest.java
package org.openapis.openapi;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

import org.openapis.openapi.models.errors.APIException;
import org.openapis.openapi.models.shared.Security;
import org.openapis.openapi.models.shared.BasicHttp;
import org.openapis.openapi.models.shared.ApiKeyAuth;
import org.openapis.openapi.models.shared.AccessToken;

public class AuthAdditionalTest {
    @Test
    void testGlobalSecurityBasicHttpSuccess() throws Exception {
        CommonHelpers.recordTest("auth-basic-http-global-option");

        // Expected to succeed since BasicHTTP takes priority over AccessToken in global security definition
        SDK s = SDK.builder()
                .security(Security.builder()
                        .basicHttp(BasicHttp.builder()
                                .username("testUser")
                                .password("testPass")
                                .build())
                        .accessToken(AccessToken.builder()
                                .accessToken("Bearer ignored")
                                .build())
                        .build())
                .build();

        var res = s.auth().globalSecurityOptionBasicHttp().call();
        assertNotNull(res);
        assertEquals(200, res.statusCode());
        assertTrue(res.basicAuth().isPresent());
        assertTrue(res.basicAuth().get().authenticated());
        assertEquals("testUser", res.basicAuth().get().user());
    }

    @Test
    void testGlobalSecurityFieldsOrdering() {
        CommonHelpers.recordTest("auth-global-security-option-fields-ordering");

        // Expected to fail since APIKeyAuth takes priority over BasicHTTP in global security definition
        SDK s = SDK.builder()
                .security(Security.builder()
                        .apiKeyAuth(ApiKeyAuth.builder()
                                .apiKeyAuth("Bearer test_api_key")
                                .build())
                        .basicHttp(BasicHttp.builder()
                                .username("testUser")
                                .password("testPass")
                                .build())
                        .build())
                .build();

        APIException ex = assertThrows(APIException.class, () -> {
            s.auth().globalSecurityOptionBasicHttp().call();
        });
        assertEquals(401, ex.code());
    }

    @Test
    void testHoistedSecurityAccessTokenFirst() throws Exception {
        CommonHelpers.recordTest("auth-hoisted-security-option-access-token-first");

        SDK s = SDK.builder()
                .security(Security.builder()
                        .apiKeyAuth(ApiKeyAuth.builder()
                                .apiKeyAuth("testApiKey")
                                .build())
                        .accessToken(AccessToken.builder()
                                .accessToken("Bearer ghp_xxxx")
                                .build())
                        .build())
                .build();

        var res = s.auth().hoistedSecurityOptionAccessTokenFirst().call();
        assertNotNull(res);
        assertEquals(200, res.statusCode());
        assertTrue(res.tokenAuthResponse().isPresent());
        assertEquals("Bearer ghp_xxxx", res.tokenAuthResponse().get().token());
    }

    @Test
    void testHoistedSecurityApiKeyFirst() throws Exception {
        CommonHelpers.recordTest("auth-hoisted-security-option-api-key-first");

        SDK s = SDK.builder()
                .security(Security.builder()
                        .apiKeyAuth(ApiKeyAuth.builder()
                                .apiKeyAuth("testApiKey")
                                .build())
                        .accessToken(AccessToken.builder()
                                .accessToken("Bearer ghp_xxxx")
                                .build())
                        .build())
                .build();

        var res = s.auth().hoistedSecurityOptionApiKeyFirst().call();
        assertNotNull(res);
        assertEquals(200, res.statusCode());
        assertTrue(res.tokenAuthResponse().isPresent());
        assertEquals("testApiKey", res.tokenAuthResponse().get().token());
    }

    @Test
    void testHoistedSecurityBasicHttpOnly() throws Exception {
        CommonHelpers.recordTest("auth-hoisted-security-option-basic-http-only");

        SDK s = SDK.builder()
                .security(Security.builder()
                        .apiKeyAuth(ApiKeyAuth.builder()
                                .apiKeyAuth("testApiKey")
                                .build())
                        .basicHttp(BasicHttp.builder()
                                .username("testUser")
                                .password("testPass")
                                .build())
                        .accessToken(AccessToken.builder()
                                .accessToken("Bearer ghp_xxxx")
                                .build())
                        .build())
                .build();

        var res = s.auth().hoistedSecurityOptionBasicHttpOnly().call();
        assertNotNull(res);
        assertEquals(200, res.statusCode());
        assertTrue(res.basicAuth().isPresent());
        assertTrue(res.basicAuth().get().authenticated());
        assertEquals("testUser", res.basicAuth().get().user());
    }

    @Test
    void testHoistedSecurityInvalidOption() {
        CommonHelpers.recordTest("auth-hoisted-security-invalid-option");

        SDK s = SDK.builder()
                .security(Security.builder()
                        .basicHttp(BasicHttp.builder()
                                .username("username")
                                .password("password")
                                .build())
                        .build())
                .build();

        APIException ex = assertThrows(APIException.class, () -> {
            s.auth().hoistedSecurityOptionAccessTokenFirst().call();
        });
        assertEquals(401, ex.code());
    }
}
